stdlib/LinearAlgebra/src/matmul.jl

```diff
@@ -106,8 +106,11 @@ julia> [1 1; 0 1] * [1 0; 1 1]
 """
 function (*)(A::AbstractMatrix, B::AbstractMatrix)
     TS = promote_op(matprod, eltype(A), eltype(B))
```

Comment:
Why not use `_init_eltype` here? What's the test you mentioned? The issue is that the `_init_eltype` mechanism exists to handle unitful matrices, where the return eltype is not obtained from promotion.

Reply:
This is the test:
https://github.com/JuliaLang/julia/blob/d90fa45c6ac5da8398e6864128579df48edbf00c/stdlib/LinearAlgebra/test/matmul.jl#L934-L941
If we use `_init_eltype`, the test that used to throw actually passes now, as `C` becomes a `6×0 Matrix{Float64}`. However, perhaps this test may be updated?

Reply:
IIRC, then `_init_eltype` falls back to `promote_op(matprod, eltype(A), eltype(B))` for eltypes other than `Number`, but for `Number`s allows to compute the unit of the product. So it should be "superior" to `promote_op`-only. And in the test, from what I understand, the issue was to make sure that the product be computed. To make sure to test code where `C` doesn't have a zero, one could modify it to

```julia
C = collect(Any, zeros(r, 0))
mul!(C, A, B)
@test size(A * B) == (r, 0)
```

Reply:
I'd suggest to do this:

```suggestion
    TS = _init_eltype(matprod, eltype(A), eltype(B))
```

In fact, we should replace all occurrences of `_init_eltype(*` by `_init_eltype(matprod` and remove the specialized method for the `*`-case. Moreover, we should have tests with unitful matrices (full and triangonal). I'm surprised that CI passes given that the eltype promotion for triangular matrices is not supposed to work without ` _init_eltype`  for unitful matrices.

Reply:
I just double-checked. It seems that things have greatly improved after v1.9 in terms of return type inference. It seems like `promote_op` works across unitful and unitless eltypes.